Level 1

13Thirteenth13 opened the pull request.

13Thirteenth13 committed:
feat: release project, added models, routes, validation middleware #1

13Thirteenth13 committed:
feat: added helmet, rate limiter #1

13Thirteenth13 committed:
added readme #1

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [ ] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [ ] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
  **Присутствуют блокировки вида `eslint-disable`.**
- [ ] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
**Ошибок не должно быть без блокировок.**
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [ ] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [ ] `GET /movies` — все сохранённые пользователем фильмы;
  **Запрос возвращает все фильмы в базе, а не фильмы текущего пользователя.**
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [ ] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [ ] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [ ] сообщение об ошибке соответсвует ее типу;
	**Не обработана ошибка дубля email при редактировании пользователя.**
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
  **Лишний `integer` для `movieId` (нет в ТЗ).**
  - [ ] Данные валидируются перед добавлением в базу.
  **Лишние ограничения для полей фильма на уровне схемы (нет в ТЗ).**
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [ ] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
**Роуты подключены отдельно.**
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [ ] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
**Обработчик описан в `app.js`.**
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
  **Сообщения прописаны вручную.**
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [ ] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
**Модуль сконфигурирован в `app.js`.**
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
- [ ] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.
**База названа `moviesdb`.**

## Количество баллов: 50

## Доп. Комментарии:
1. **Лишний `crash-test`.**
2. **При создании фильма можно обойтись без паровоза деструктуризации:**
```js
Movie.create({ ...req.body, owner: req.user._id })
```
3. **Лишний `getUsers`.**
4. **Перемудренная логика отправки документа. Просто отправляйте обновленную версию из базы.**
![2022-12-30_04-15-22](https://user-images.githubusercontent.com/86304001/210016185-7def2422-436b-4578-91e3-fc2bf7754bb7.png)
5. **Бессмысленная проверка на `celebrate`. Ошибки `celebrate` обрабатываются отдельно.**
![2022-12-30_04-18-30](https://user-images.githubusercontent.com/86304001/210017151-70a35794-7f47-4edf-b794-7a86eaa3879c.png)

13Thirteenth13 committed:
fix: 1st review fixes #1

Feat: error middleware; error messeges in ./error/index.js; added index.js in routes. Fixes: eslint-disable; GET /movies; createMovie; movie model; db name. Removed getUsers, crash-test; integer for movieId.

13Thirteenth13 commented:
🠕 Описание правок и исправлений в боди коммита 🠕

(P.S. "Не обработана ошибка дубля email при редактировании пользователя." - тоже исправлено)

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [x] сообщение об ошибке соответсвует ее типу;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
  - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
  - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [x] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [ ] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
**Модуль сконфигурирован в `app.js`.**
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 99
